### ISSUE TITLE
Appropriate Recursion Combinator

### ISSUE THREAD

**Issue by PetarMax** [runtimeverification/plutus-core-semantics#255]
I've done an initial investigation and, for example, all `auction` examples in `benchmark-validation-examples` seem to use the following combinator for recursion:

```
(lam
  f_0
  [
    (lam z_0 [ z_0 z_0 ])
    (lam x_0 (lam y_0 [ [ f_0 [ x_0 x_0 ] ] y_0 ]))
  ]
)
```

which is `beta`-equal to

```
(lam
  f_0
  [
    (lam x_0 (lam y_0 [ [ f_0 [ x_0 x_0 ] ] y_0 ]))
    (lam x_0 (lam y_0 [ [ f_0 [ x_0 x_0 ] ] y_0 ]))
  ]
)
```

which looks like, but is not quite, the `Z` combinator we are using:

```
(lam 
  f_0 
  [
    (lam x_0 [ f_0 (lam y_0 [ x_0 x_0 y_0 ])]) 
    (lam x_0 [ f_0 (lam y_0 [ x_0 x_0 y_0 ])])
  ]
)
```

However, both this combinator (I'll call it `X`) and `Z` respect the following:

```
Xgv => g(Xg)v
Zgv => g(Zg)v
```

and so `X` and `Z` are `beta`-equivalent to each other, which is good news, I guess. 

Could we confirm with the client that this is the only recursive combinator used? In any case, I can adapt the examples to use it instead of `Z`.

**Comment by ChristianoBraga** [comment]
Great Petar! There are also the factoria and fibonacci examples in `tests/textual/simple`. I can ask but you can also post your question directly at the `plutus-semantics-rv` channel. Let me know.

**Comment by PetarMax** [comment]
Yes, that combinator is almost the same as the one in the other examples. They seem to be doing

```
(lam x [x x]) y
```

deliberately instead of directly going for

```
[y y]
```

and I would like to understand why. I conjecture that it might be computationally more efficient, because `y` is then stored in the environment in `x` instead of being duplicated in the term itself.
